server/gc_service.go

```diff
@@ -208,57 +426,38 @@ func (s *GrpcServer) GetAllGCSafePointV2(ctx context.Context, request *pdpb.GetA
 		return rsp.(*pdpb.GetAllGCSafePointV2Response), err
 	}
 
-	startkey := keypath.GCSafePointV2Prefix()
-	endkey := clientv3.GetPrefixRangeEnd(startkey)
-	values, revision, err := s.loadRangeFromEtcd(startkey, endkey)
-
-	gcSafePoints := make([]*pdpb.GCSafePointV2, 0, len(values))
-	for _, value := range values {
-		jsonGcSafePoint := &endpoint.GCSafePointV2{}
-		if err = json.Unmarshal([]byte(value), jsonGcSafePoint); err != nil {
-			return nil, errs.ErrJSONUnmarshal.Wrap(err).GenWithStackByCause()
-		}
-		gcSafePoint := &pdpb.GCSafePointV2{
-			KeyspaceId:  jsonGcSafePoint.KeyspaceID,
-			GcSafePoint: jsonGcSafePoint.SafePoint,
-		}
-		log.Debug("get all gc safe point v2",
-			zap.Uint32("keyspace-id", jsonGcSafePoint.KeyspaceID),
-			zap.Uint64("gc-safe-point", jsonGcSafePoint.SafePoint))
-		gcSafePoints = append(gcSafePoints, gcSafePoint)
+	rc := s.GetRaftCluster()
+	if rc == nil {
+		return &pdpb.GetAllGCSafePointV2Response{Header: notBootstrappedHeader()}, nil
 	}
 
+	gcStates, err := s.gcStateManager.GetAllKeyspacesGCStates()
 	if err != nil {
 		return &pdpb.GetAllGCSafePointV2Response{
 			Header: wrapErrorToHeader(pdpb.ErrorType_UNKNOWN, err.Error()),
-		}, err
+		}, nil
+	}
+
+	gcSafePoints := make([]*pdpb.GCSafePointV2, 0, len(gcStates))
+	for _, gcState := range gcStates {
+		if gcState.KeyspaceID == constant.NullKeyspaceID {
+			// The original GetAllGCSafePointV2 API doesn't return the null keyspace. To keep the behavior, we
+			// still exclude it.
+			continue
+		}
+		gcSafePoints = append(gcSafePoints, &pdpb.GCSafePointV2{
+			KeyspaceId:  gcState.KeyspaceID,
+			GcSafePoint: gcState.GCSafePoint,
+		})
 	}
 
 	return &pdpb.GetAllGCSafePointV2Response{
 		Header:       wrapHeader(),
 		GcSafePoints: gcSafePoints,
-		Revision:     revision,
+		Revision:     0,
```

Comment:
Is the revision not used anymore?

Reply:
Actually I don't know when it's useful, but now this API cannot provide this field anymore.
cc @ystaticy do you know the purpose of this field?